prov/gni/src/gnix_msg.c

```diff
@@ -1651,3 +2404,441 @@ ssize_t _gnix_send(struct gnix_fid_ep *ep, uint64_t loc_addr, size_t len,
+
+						for (i--; i >= 0; i--) {
+							fi_close(&req->msg.recv_md[i]->mr_fid.fid);
+						}
+
+						return ret;
+					}
+
+					req->msg.recv_md[i] = container_of(
+						(void *) auto_mr,
+						struct gnix_fid_mem_desc,
+						mr_fid);
+
+					req->msg.recv_info[i].recv_addr = (uint64_t) iov[i].iov_base;
+					req->msg.recv_info[i].recv_len = iov[i].iov_len;
+					req->msg.recv_info[i].mem_hndl =
+						req->msg.recv_md[i]->mem_hndl;
+
+					GNIX_DEBUG(FI_LOG_EP_DATA, "auto-reg MR"
+						   ": %p\n",
+						   req->msg.recv_md[i]);
+
+				}
+
+				req->msg.send_flags |= FI_LOCAL_MR;
+			} else {	/* User registered their memory */
+				for (i = 0; i < count; i++) {
+					if (!desc[i]) {
+						GNIX_FATAL(FI_LOG_EP_DATA,
+							   "invalid memory reg"
+							   "istration (%p).\n",
+							   desc[i]);
+					}
+
+					req->msg.recv_md[i] =
+						container_of(desc[i],
+						struct gnix_fid_mem_desc,
+						mr_fid);
+
+					req->msg.recv_info[i].mem_hndl =
+						req->msg.recv_md[i]->mem_hndl;
+				}
+			}
+
+			ret = _gnix_vc_queue_work_req(req);
+		} else {
+			for (i = 0; i < count; i++) {
+				req->msg.recv_info[i].recv_addr = (uint64_t) iov[i].iov_base;
+				req->msg.recv_info[i].recv_len = iov[i].iov_len;
+			}
+
+			/*
+			 * This request is associate with a regular eager smsg,
+			 * the rndzv threshold on the sender was not reached or
+			 * exceeded.
+			 */
+			__gnix_msg_unpack_data_into_iov(req->msg.recv_info,
+							count,
+							req->msg.send_info[0].send_addr,
+							req->msg.send_info[0].send_len);
+
+			__gnix_msg_recv_completion(ep, req);
+			_gnix_fr_free(ep, req);
+		}
+	} else {
+		/* if peek/claim/discard, we didn't find what we
+		 * were looking for, return FI_ENOMSG
+		 */
+		if (match_flags) {
+			__recv_err(ep, context, flags, cum_len,
+				   (void *) iov, 0, tag, cum_len, FI_ENOMSG,
+				   FI_ENOMSG, NULL);
+
+			/* if handling trecvmsg flags, return here
+			 * Never post a receive request from this type of
+			 * context
+			 */
+			ret = FI_SUCCESS;
+			goto pdc_exit;
+		}
+
+		/*
+		 * No matching requests found, create a new one and enqueue
+		 * it in the posted queue.
+		 */
+		req = _gnix_fr_alloc(ep);
+		if (req == NULL) {
+			ret = -FI_EAGAIN;
+			goto err;
+		}
+
+		GNIX_DEBUG(FI_LOG_EP_DATA, "EXPECTED, req = %p\n", req);
+
+		if (tagged) {
+			req->type = GNIX_FAB_RQ_TRECVV;
+		} else {
+			req->type = GNIX_FAB_RQ_RECVV;
+		}
+
+		if ((flags & GNIX_SUPPRESS_COMPLETION) ||
+		    (ep->recv_selective_completion &&
+		     !(flags & FI_COMPLETION))) {
+			req->msg.recv_flags &= ~FI_COMPLETION;
+		} else {
+			req->msg.recv_flags |= FI_COMPLETION;
+		}
+
+		req->addr = gnix_addr;
+		req->gnix_ep = ep;
+		req->user_context = context;
+		req->flags = 0;
+
+		for (i = 0; i < count; i++) {
+			req->msg.recv_info[i].recv_addr = (uint64_t) iov[i].iov_base;
+			req->msg.recv_info[i].recv_len = iov[i].iov_len;
+			/* req->msg.recv_info[i].mem_hndl ^= req->msg.recv_info[i].mem_hndl; */
```

Comment:
> What's this left from?

Testing the changes from changing the fabric request fields to be arrays of length GNIX_MAX_IOV_LIMIT.